### ISSUE TITLE
Reintroduce logic for admin_level 4 boundaries as part of large countries

### ISSUE THREAD

**Issue by pnorman** [gravitystorm/openstreetmap-carto#2280]
We reverted #1938 for performance reasons in #1987. Now that we have a set of suggested indexes, we can reconsider it if the performance with indexes is acceptable.

**Comment by nebulon42** [comment]
I will try to test this locally with data filtered by https://github.com/gmgeo/osm-carto-lowzoom.

**Comment by pnorman** [comment]
the performance question is will planet_osm_polygon_way_area_z6 be acceptable for performance or do we need to add an index on  `boundary IS NOT NULL` or `boundary = 'administrative' AND admin_level IN ('0', '1', '2', '3', '4')`

**Comment by nebulon42** [comment]
I'm not sure if my setup with low-zoom global data is representative or not. Now the database contains all data up to z8 but still misses quite something. Since I'm not able to load the full planet this may not help much.

> boundary = 'administrative' AND admin_level IN ('0', '1', '2', '3', '4')

This index is already suggested in the index file.

I have taken the query you referenced in https://github.com/gravitystorm/openstreetmap-carto/pull/1938#issuecomment-160944169 and tested it with and without additional indexes.

Query with suggested indexes: https://explain.depesz.com/s/PYW
"Planning time: 3.472 ms"
"Execution time: 185195.312 ms"

Query with only standard indexes: https://explain.depesz.com/s/SiCRy
"Planning time: 3.191 ms"
"Execution time: 190091.128 ms"

The runtimes and the query plans are quite similar. If it helps I'm running PostgreSQL 9.5.
If I have to do something different I'm happy for suggestions.

**Comment by pnorman** [comment]
> I'm not sure if my setup with low-zoom global data is representative or not.

No. The majority of the polygon table is normally buildings. This makes a bitmap heap scan possible to do on your setup, particularly if you're using mechanical drives. 3 minutes is an acceptable time, but I'm not confident it would be the same in production.

That being said, with the right indexes it shouldn't do a bitmap heap scan so the query plan indicates a problem.

**Comment by pnorman** [comment]
For a full planet it takes [about 250s](https://explain.depesz.com/s/YtR), which is acceptable, but improvements can be made.

Looking at the query, it's doing a BitmapOr on two identical nodes. A query that gives a better plan is

```sql
SELECT ST_AsBinary("way") AS geom,"admin_level" FROM (SELECT
    way,
    admin_level
  FROM planet_osm_roads o
  WHERE boundary = 'administrative'
    AND o.admin_level IN ('0', '1', '2', '3', '4') -- All possible admin boundaries to render
    AND (o.admin_level IN ('0', '1', '2') -- admin boundary is low enough level to unconditionally render
      OR EXISTS -- Or is part of a large country
        (SELECT 1
          FROM planet_osm_polygon i
          WHERE i.boundary = 'administrative'
            AND i.admin_level = '2'
            AND i.way_area > 9e+12
            AND ST_Within(o.way, i.way)
            AND i.osm_id < 0
        )
      )
    AND o.osm_id < 0
  ORDER BY admin_level DESC
  ) AS admin_very_low_zoom WHERE "way" && ST_SetSRID('BOX3D(-1252344.271424999 -1252344.271424999,20037508 20037508)'::box3d, 900913);
```

Which [takes 144s](https://explain.depesz.com/s/ZWll).

Adding a `way && !bbox!` condition to both didn't help.

On the index side, we can make some improvements. Layers that query planet_osm_polygon on some boundary is not null value are

1. `nature-reserve-boundaries` with `(boundary = 'national_park' OR leisure = 'nature_reserve') AND building IS NULL`
2. `placenames-large` with `boundary = 'administrative' AND admin_level IN ('2', '4') AND name IS NOT NULL`
3. `text-poly-low-zoom` with an OR between it, landuse, natural, place, and leisure and an `AND building IS NOT NULL`
4. `text-poly` as part of a monster conditional
5. `admin-text` with `boundary = 'administrative' AND admin_level IN ('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10') AND name IS NOT NULL`
6. `nature-reserve-text` with `(boundary = 'national_park' OR leisure = 'nature_reserve') AND name IS NOT NULL`

2 and 5 are related and an index could help them too. The obvious index conditions are

1. `WHERE boundary IS NOT NULL`
2. `WHERE boundary = 'administrative'`
3. `WHERE boundary = 'administrative' AND admin_level IN ('0', '1', '2', '3', '4')`
4. `WHERE boundary IS NOT NULL OR leisure = 'nature_reserve'`

Once some indexes build, I'll be able to test them.

**Comment by pnorman** [comment]
For reference, planet_osm_polygon_way_area_z6 is 1175 MB 

> WHERE boundary IS NOT NULL

Doesn't help the very low zoom query much. Index is 62 MB

> `WHERE boundary = 'administrative'`

Statistically the same as `boundary IS NOT NULL`, not evaluated.

> `WHERE boundary = 'administrative' AND admin_level IN ('0', '1', '2', '3', '4')`.

Also doesn't help much. Index is 1280 kB.

I think the slowness here is not the index, but the number of pages it has to pull from disk. The way_area_z6 index is probably good enough for the query.

**Comment by jeisenbe** [comment]
Re: https://github.com/gravitystorm/openstreetmap-carto/issues/2280#issuecomment-242544767

> will planet_osm_polygon_way_area_z6 be acceptable

Did we decide that the answer is "yes", or "not sure" still?

I imagine that if we can replace Natural Earth with preprocessed OpenStreetMap data, we could choose to include admin_level=4 boundaries for large countries in the same data, and use that at z4 and perhaps z5? See #2172
